File: cpp/log-levels/log_levels.cpp
#include <string>

namespace log_line {
    std::string message(const std::string& log_line) {
        int msg_start = log_line.find(": ") + 2;
        return log_line.substr(msg_start);
    }

    std::string log_level(const std::string& log_line) {
        int level_end = log_line.find("]") -1;
        return log_line.substr(1, level_end);
    }

    std::string reformat(const std::string log_line) {
        return message(log_line) + " (" + log_level(log_line) + ")";
    }
} // namespace log_line

File: r/log-levels/1/log-levels.R
# First solution sticks to base R

split_msg <- function(msg) {
  strsplit(msg, ": ")[[1]]
}

message <- function(msg) {
  trimws(split_msg(msg)[2])
}

log_level <- function(msg) {
  raw <- tolower(split_msg(msg)[1])
  len <- nchar(raw)
  substr(raw, 2, len - 1)
}

reformat <- function(msg) {
  paste(message(msg), " (", log_level(msg), ")", sep = "")
}
